### PR TITLE
oidc client auth provider: cache OpenID Connect clients to prevent reinitialization

### LINKED ISSUE #37876: kubectl auth providers are repeatedly reinitialized
When enabling the OpenID Connect auth provider we found that auth providers are initialized almost every time kubectl needs to make a new request rather than reusing the initialized transport. @mlbiam noticed this as well as it causes kubectl to become excruciatingly slow.

Will follow up with a more detailed bug report.

cc @cjcullen @jlowdermilk (don't know who else to CC)

### PR DESCRIPTION
**Pull request by ericchiang**
Prior to this change, the OIDC client auth provider synced the OIDC provider metadata with every transport initialization. When used with kubectl, the provider can be reinitialize dozens of times, each triggering a round trip to the OIDC auth server. The overhead makes this feature largely unusable, adding several seconds to every kubectl invocation, and we observe many users simply abandoning it for other workarounds.

This fix prevents the excessive round trips by keeping a global cache of clients similarly to the TLS transport logic in client-go at the result of a significant speedup. The caching also fixes data races when a token is refreshed.

closes #37876

cc @kubernetes/sig-auth @liggitt @jsloyer @mlbiam @philips 

```release-note
Caching added to the OIDC client auth plugin to fix races and reduce the time kubectl commands using this plugin take by several seconds.
```

### CODE REVIEW

**Comment by philips** [comment]
Manually tested and I can happily confirm it reduces the number of GET requests to the openid-configuration URL from 60+ to 1:

```
$ kubectl get pods  -v=9 2>&1 | grep well-known
I1205 20:14:38.118650    7143 round_trippers.go:393] curl -k -v -XGET  https://tec.a.ifup.org:32000/identity/.well-known/openid-configuration
I1205 20:14:38.311009    7143 round_trippers.go:412] GET https://tec.a.ifup.org:32000/identity/.well-known/openid-configuration 200 OK in 192 milliseconds
```

**Comment by jsloyer** [comment]
@philips.  I have seen some requests taking 15 seconds with an OIDC provider, what did you observe an API call through kubectl taking with this change?

**Comment by philips** [comment]
@jsloyer `kubectl get pods` returns in 0.5s. Cluster: AWS us-west-1 accessed from residential Comcast running Tectonic and Dex v2 on cluster backed by TPRs.

```
$ time kubectl get pods
NAME                            READY     STATUS    RESTARTS   AGE
etcd-operator-217127005-1pi7l   1/1       Running   0          2d

real	0m0.503s
user	0m0.091s
sys	0m0.020s
```

**Comment by jsloyer** [comment]
This is amazing, this is a huge speed up.  I built and tested a kubectl client with this change

File: plugin/pkg/client/auth/oidc/oidc.go [hunk 1]
```diff
@@ -75,57 +110,61 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
 		return nil, fmt.Errorf("Must provide %s", cfgClientSecret)
 	}
 
-	var certAuthData []byte
-	var err error
-	if cfg[cfgCertificateAuthorityData] != "" {
-		certAuthData, err = base64.StdEncoding.DecodeString(cfg[cfgCertificateAuthorityData])
-		if err != nil {
-			return nil, err
+	client, ok := cache.getClient(issuer, clientID, clientSecret)
```
Comment (liggitt) [inline plugin/pkg/client/auth/oidc/oidc.go]: return early and unnest?
```
if ok {
  return client, nil
}
```

File: plugin/pkg/client/auth/oidc/oidc.go [hunk 2]
```diff
@@ -75,57 +110,61 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
...
+				Secret: clientSecret,
+			},
+			ProviderConfig: providerCfg,
+			Scope:          append(scopes, oidc.DefaultScope...),
+		}
+		client, err = oidc.NewClient(oidcCfg)
+		if err != nil {
+			return nil, fmt.Errorf("error creating OIDC Client: %v", err)
+		}
+		cache.setClient(issuer, clientID, clientSecret, client)
```
Comment (liggitt) [inline plugin/pkg/client/auth/oidc/oidc.go]: still have the potential to construct this multiple times if newOIDCAuthProvider is called from multiple threads... are we ok with that?

**Comment by liggitt** [comment]
a few nits, LGTM

File: plugin/pkg/client/auth/oidc/oidc.go (hunk 1 above)
Comment (ericchiang) [inline plugin/pkg/client/auth/oidc/oidc.go]: There's just enough logic after the if statement to make that annoying. Let me see if I can split it out to a helper or something.

**Comment by ericchiang** [comment]
Quick update on this PR. I've been diving a bit more into the implementation and it seems like there are some bigger issues here. For example the client doesn't sync around requesting refresh tokens and updating the kubeconfig. If it's used concurrently there's a very basic race there. Also the returned transport seems to be doing retries[0], which makes some bold assumptions about the server not consuming POST body data, and means we're not closing response bodies.

Trying to clean this stuff up along with this change.

[0] https://github.com/kubernetes/kubernetes/blob/407d8b7c190aef1e1774beb21fa1f5e67e13e40d/plugin/pkg/client/auth/oidc/oidc.go#L176

**Comment by ericchiang** [comment]
@liggitt updated though this became a bit bigger of a PR.

Highlights:

* Cache OpenID Connect clients to prevent reinitialization
* Don't retry requests in the http.RoundTripper.
  * Don't rely on the server not reading POST bodies.
  * Don't leak response body FDs.
  * Formerly ignored any throttling requests by the server.
* Determine if the id token's expired by inspecting it.
  * Similar to logic in golang.org/x/oauth2
* Synchronize around refreshing tokens and persisting the new config.

**Comment by ericchiang** [comment]
@k8s-bot bazel test this

**Comment by ericchiang** [comment]
@k8s-bot verify test this

**Comment by ericchiang** [comment]
cc @yifan-gu

File: plugin/pkg/client/auth/oidc/oidc.go [hunk 3]
```diff
@@ -44,21 +44,63 @@ const (
...
+}
+
+// setClient attempts to put the client in the cache but may return any clients
+// with the same keys set before. This is so there's only ever one client for a provider.
+func (c *clientCache) setClient(issuer, clientID, clientSecret string, client *oidcAuthProvider) *oidcAuthProvider {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	key := cacheKey{issuer, clientID, clientSecret}
+
+	if oldClient, ok := c.cache[key]; ok {
```
Comment (yifan-gu) [inline plugin/pkg/client/auth/oidc/oidc.go]: why we want to return the old client?
if we want to use the old client when it's there we can always do getClient() first.

In fact, the newOIDCAuthProvider() function below, the setClient() is always returning the latest client.

**Review by yifan-gu — COMMENTED** [review]
@ericchiang Just some nits, LGTM

File: plugin/pkg/client/auth/oidc/oidc.go (hunk 3 above)
Comment (ericchiang) [inline plugin/pkg/client/auth/oidc/oidc.go]: Let me add a comment.

File: plugin/pkg/client/auth/oidc/oidc.go [hunk 4]
```diff
@@ -112,146 +159,134 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
...
-		return fmt.Errorf("token already expired at: %v", exp)
+	// shallow copy of the struct
+	r2 := new(http.Request)
+	*r2 = *req
+	// deep copy of the Header
+	r2.Header = make(http.Header)
+	for k, s := range req.Header {
+		r2.Header[k] = s
 	}
+	r2.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
```
Comment (ericchiang) [inline plugin/pkg/client/auth/oidc/oidc.go]: To comply with the RoundTripper interface

> RoundTrip should not modify the request...

https://golang.org/pkg/net/http/#RoundTripper

If we don't copy deepcopy the header then we're adding a bearer toke to the original header, modifying the original request.

Comment (ericchiang) [inline plugin/pkg/client/auth/oidc/oidc.go]: Same logic as the oauth2 package https://github.com/golang/oauth2/blob/314dd2c0bf3ebd592ec0d20847d27e79d0dbe8dd/transport.go#L94-L106

**Comment by ericchiang** [comment]
@yifan-gu added some inline comments

File: plugin/pkg/client/auth/oidc/oidc.go (hunk 4 above)
Comment (yifan-gu) [inline plugin/pkg/client/auth/oidc/oidc.go]: @ericchiang Gotcha.

File: plugin/pkg/client/auth/oidc/oidc.go (hunk 3 above)
Comment (yifan-gu) [inline plugin/pkg/client/auth/oidc/oidc.go]: @ericchiang I see, it's a testAndSet operation, maybe rename the func?

**Comment by yifan-gu** [comment]
LGTM

**Comment by ericchiang** [comment]
@k8s-bot bazel test this

**Comment by ericchiang** [comment]
cc @kubernetes/sig-auth-misc anyone else interested in reviewing this? Without this `kubectl` is relatively unusable with this auth plugin if you're not sitting next to your cluster :)

File: plugin/pkg/client/auth/oidc/oidc.go [hunk 5]
```diff
@@ -112,146 +164,135 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
...
 type oidcAuthProvider struct {
-	refresher      *idTokenRefresher
-	initialIDToken jose.JWT
+	// Interface rather than a raw *oidc.Client for testing.
+	client OIDCClient
+
+	// Stubbed out for testing.
+	now func() time.Time
+
+	// Mutex guards presisting to the kubeconfig file and allows synchronized
```
Comment (liggitt) [inline plugin/pkg/client/auth/oidc/oidc.go]: persisting

File: plugin/pkg/client/auth/oidc/oidc.go [hunk 6]
```diff
@@ -112,146 +164,135 @@ func newOIDCAuthProvider(_ string, cfg map[string]string, persister restclient.A
...
-		r.cfg[cfgRefreshToken] = tokens.RefreshToken
+		newCfg[cfgRefreshToken] = tokens.RefreshToken
 	}
-	r.cfg[cfgIDToken] = jwt.Encode()
 
-	err = r.persister.Persist(r.cfg)
-	if err != nil {
-		return jose.JWT{}, fmt.Errorf("could not perist new tokens: %v", err)
+	idToken := jwt.Encode()
+	newCfg[cfgIDToken] = idToken
```
Comment (liggitt) [inline plugin/pkg/client/auth/oidc/oidc.go]: should this just be `newCfg[cfgIDToken] = tokens.IDToken`?

**Comment by liggitt** [comment]
just a couple small things, LGTM

**Comment by ericchiang** [comment]
@k8s-bot bazel test this

**Comment by ericchiang** [comment]
@k8s-bot bazel test this

**Comment by philips** [comment]
@k8s-bot bazel test this

Filed an issue about the empty result https://github.com/kubernetes/test-infra/issues/1465

**Comment by philips** [comment]
@k8s-bot etcd3 test this

**Comment by philips** [comment]
OK, tests finally went green. LGTM at this point but I am not an owner.

**Comment by liggitt** [comment]
LGTM

**Comment by ericchiang** [comment]
Anyone know how to retrigger the kops build?

**Comment by ericchiang** [comment]
@k8s-bot kops test this

**Pull request merged**